frontend/spec/features/coupon_code_spec.rb

```diff
@@ -32,50 +32,106 @@ def create_basic_coupon_promotion(code)
-      end
+      context "as guest without registration" do
+        before do
+          visit spree.root_path
+          click_link "RoR Mug"
+          click_button "add-to-cart-button"
+          click_button "Checkout"
+          fill_in "order_email", with: "spree@example.com"
+          fill_in "First Name", with: "John"
+          fill_in "Last Name", with: "Smith"
+          fill_in "Street Address", with: "1 John Street"
+          fill_in "City", with: "City of John"
+          fill_in "Zip", with: "01337"
+          select country.name, from: "Country"
+          select state.name, from: "order[bill_address_attributes][state_id]"
+          fill_in "Phone", with: "555-555-5555"
+
+          # To shipping method screen
+          click_button "Save and Continue"
+          # To payment screen
+          click_button "Save and Continue"
+        end
 
-      it "can enter an invalid coupon code, then a real one" do
-        fill_in "order_coupon_code", with: "coupon_codes_rule_man"
-        click_button "Save and Continue"
-        expect(page).to have_content(Spree.t(:coupon_code_not_found))
-        fill_in "order_coupon_code", with: "onetwo"
-        click_button "Save and Continue"
-        expect(page).to have_content("Promotion (Onetwo)   -$10.00")
-      end
+        it "informs about an invalid coupon code" do
+          fill_in "order_coupon_code", with: "coupon_codes_rule_man"
+          click_button "Save and Continue"
+          expect(page).to have_content(Spree.t(:coupon_code_not_found))
+        end
 
-      context "with a promotion" do
-        it "applies a promotion to an order" do
+        it "can enter an invalid coupon code, then a real one" do
+          fill_in "order_coupon_code", with: "coupon_codes_rule_man"
+          click_button "Save and Continue"
+          expect(page).to have_content(Spree.t(:coupon_code_not_found))
           fill_in "order_coupon_code", with: "onetwo"
           click_button "Save and Continue"
           expect(page).to have_content("Promotion (Onetwo)   -$10.00")
         end
+
+        context "with a promotion" do
+          it "applies a promotion to an order" do
+            fill_in "order_coupon_code", with: "onetwo"
+            click_button "Save and Continue"
+            expect(page).to have_content("Promotion (Onetwo)   -$10.00")
+          end
+        end
+      end
+
+      context 'as logged user' do
+        let!(:user) do
+          create(:user, bill_address: create(:address), ship_address: create(:address))
+        end
+
+        let!(:order) do
+          order = Spree::Order.create!(
+            email: "spree@example.com",
+            store: Spree::Store.first || FactoryGirl.create(:store)
+          )
+
+          order.reload
+          order.user = user
+          order.recalculate
+          order
+        end
+
+        before do
+          allow_any_instance_of(Spree::CheckoutController).to receive_messages(current_order: order)
```

Comment:
I don't like the use of stubs here. The feature specs are intended to interact with the site as a user would. Here we create an order for them and stub things rather than browsing through the site to add stuff to a cart.

Reply:
yes, you are right. I changed the code to avoid stubbing the current order